taffy/main.go

```diff
@@ -0,0 +1,111 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/github.com/flynn/go-docopt"
+	"github.com/flynn/flynn/pkg/version"
+)
+
+func init() {
+	log.SetFlags(0)
+}
+
+const DEST = "/tmp/app"
+
+func parsePairs(args *docopt.Args, str string) (map[string]string, error) {
+	pairs := args.All[str].([]string)
+	item := make(map[string]string, len(pairs))
+	for _, s := range pairs {
+		v := strings.SplitN(s, "=", 2)
+		if len(v) != 2 {
+			return nil, fmt.Errorf("invalid var format: %q", s)
+		}
+		item[v[0]] = v[1]
+	}
+	return item, nil
+}
+
+func main() {
+	usage := `
+Usage: taffy <app> <repo> <branch> <rev> [-e <var>=<val>]... [-m <key>=<val>]...
+
+Options:
+	-e,--env <var>=<val>
+	-m,--meta <key>=<val>
+`[1:]
+	args, _ := docopt.Parse(usage, nil, true, version.String(), false)
+
+	app := args.String["<app>"]
+	repo := args.String["<repo>"]
+	branch := args.String["<branch>"]
+	rev := args.String["<rev>"]
+
+	meta := map[string]string{
+		"git":       "true",
+		"clone_url": repo,
+		"branch":    branch,
+		"rev":       rev,
+		"taffy_job": os.Getenv("FLYNN_JOB_ID"),
+	}
+
+	env, err := parsePairs(args, "--env")
+	if err != nil {
+		log.Fatal(err)
+	}
+	m, err := parsePairs(args, "--meta")
+	if err != nil {
+		log.Fatal(err)
+	}
+	for k, v := range m {
+		meta[k] = v
+	}
+
+	if err := cloneRepo(repo, branch); err != nil {
+		log.Fatal(err)
+	}
+	if err := runReceiver(app, rev, env, meta); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func cloneRepo(repo, branch string) error {
+	cmd := exec.Command("git", "clone", "--depth=50", fmt.Sprintf("--branch=%s", branch), repo, DEST)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+func gitArchive(rev string, stdout io.Writer) error {
+	cmd := exec.Command("git", "-C", DEST, "archive", rev)
+	cmd.Stdout = stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+func runReceiver(app, rev string, env, meta map[string]string) error {
+	var buf bytes.Buffer
+	args := make([]string, 0, len(env)+len(meta)+2)
+	args = append(args, app)
+	args = append(args, rev)
+	for name, m := range map[string]map[string]string{"--env": env, "--meta": meta} {
+		for k, v := range m {
+			args = append(args, name)
+			args = append(args, fmt.Sprintf("%s=%s", k, v))
+		}
+	}
+	cmd := exec.Command("/bin/flynn-receiver", args...)
+	cmd.Stdin = &buf
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := gitArchive(rev, &buf); err != nil {
+		return err
+	}
```

Comment:
This should be run in a goroutine with an `io.Pipe` instead of buffering the entire tarball in memory.